### PR TITLE
updated Warg and refactored

### PR DESCRIPTION
**Pull request by calvinrp**

### CODE REVIEW

File: crates/wasm-pkg-loader/src/source/warg.rs [hunk 1]
```diff
@@ -88,22 +94,26 @@ impl PackageSource for WargSource {
...
-            .boxed())
+        let package_name = package.try_into()?;
+
+        // warg client already validated the digest
+        let download = self
+            .client
+            .download_exact(&package_name, &release.version)
+            .await?;
+        let file = tokio::fs::File::open(download.path).await?;
+        Ok(ReaderStream::new(file).map_err(Into::into).boxed())
```
Comment (lann) [inline crates/wasm-pkg-loader/src/source/warg.rs:116]: Is this change necessary? It isn't always desirable to write to disk...

File: crates/wasm-pkg-loader/src/config.rs [hunk 2]
```diff
@@ -44,6 +44,10 @@ impl ClientConfig {
         self
     }
 
+    pub fn get_default_registry(&self) -> Option<&str> {
```
Comment (lann) [inline crates/wasm-pkg-loader/src/config.rs:47]: `get_` is discouraged: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: crates/wasm-pkg-loader/src/main.rs [hunk 3]
```diff
@@ -20,11 +22,30 @@ async fn main() -> anyhow::Result<()> {
...
+                _ => {}
+            };
         }
+
+        // if not configured and not using Warg home URL as default registry,
+        // then set the `wasi` default registry
+        if is_missing_config && config.get_default_registry().is_none() {
+            config.namespace_registry("wasi", "bytecodealliance.org");
+        };
+
```
Comment (lann) [inline crates/wasm-pkg-loader/src/main.rs:48]: Let's discuss this in https://github.com/bytecodealliance/wasm-pkg-tools/discussions/3

File: crates/wasm-pkg-loader/src/source/warg.rs [hunk 4]
```diff
@@ -61,7 +62,13 @@ impl PackageSource for WargSource {
...
             .releases()
-            .map(|release| release.version.clone())
+            .filter_map(|release| {
+                if release.yanked() {
+                    None
+                } else {
+                    Some(release.version.clone())
+                }
+            })
             .collect())
```
Comment (lann) [inline crates/wasm-pkg-loader/src/source/warg.rs:72]: I wonder if this should be configurable, e.g. an `include_yanked` options.

File: crates/wasm-pkg-loader/src/config.rs (hunk 2 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/config.rs:47]: Yeah, there was a name conflict. I'll refactor the other names to have `set_` prefix.

File: crates/wasm-pkg-loader/src/source/warg.rs (hunk 4 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs:72]: Or maybe label yanked in the output list?

Comment (lann) [reply crates/wasm-pkg-loader/src/source/warg.rs:72]: Yeah but it also might make sense to allow (attempting to) fetching the actual artifact too.

File: crates/wasm-pkg-loader/src/source/warg.rs (hunk 1 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: I'll refactor to ignore the Warg cache

Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: Is it a problem that Warg caches the package logs? Or do you need entirely in-memory?

Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: The solution feels like adding an `InMemoryClient` instead of using `FileSystemClient` in `warg_client`. Thoughts?

Comment (lann) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: There is no (durable) writable disk available in this scenario; caching is handled at a higher level. You're probably right that a separate `InMemoryClient` would make sense as well, though its a little less important for the relatively small registry metadata.

Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: Should this crate CLI never use file system Warg download cache? Or should it be configuration?

File: crates/wasm-pkg-loader/src/main.rs (hunk 3 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/main.rs:48]: Just caught up on that thread. I think this is in inline with what was discussed there.

If global `wasm-pkg-tools` config file is not defined but the global `warg` config file is defined, then it falls back to using the `warg` config given this tool can use the `warg_client`.

If there is no global config for either, it uses `bytecodealliance.org`.

File: crates/wasm-pkg-loader/src/source/warg.rs (hunk 1 above)
Comment (lann) [reply crates/wasm-pkg-loader/src/source/warg.rs:116]: It just needs some path that doesn't require the entire package content to be written to disk.

File: crates/wasm-pkg-loader/src/config.rs (hunk 2 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/config.rs]: Refactored, let me know if that is OK now.

File: crates/wasm-pkg-loader/src/main.rs (hunk 3 above)
Comment (lann) [reply crates/wasm-pkg-loader/src/main.rs]: Having two different sources for "default registry" interacting this way seems potentially confusing. It might make sense to consolidate config between `warg` and this library a bit to end up with a similar UX; that's what I'm trying to hash out in the discussion thread.

File: crates/wasm-pkg-loader/src/source/warg.rs (hunk 4 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs]: Pushed an update will list all versions and mark yanked ones.

File: crates/wasm-pkg-loader/src/source/warg.rs (hunk 1 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/source/warg.rs]: Fixed with the updated Warg 0.6.0

File: crates/wasm-pkg-loader/src/main.rs (hunk 3 above)
Comment (calvinrp) [reply crates/wasm-pkg-loader/src/main.rs]: Complete agreement. Do you think this implementation makes sense for this PR? And then we work on the rest in follow on PRs?

Comment (lann) [reply crates/wasm-pkg-loader/src/main.rs]: I don't think we want to land this kind of fundamental config logic without discussion: https://github.com/bytecodealliance/wasm-pkg-tools/discussions/10

Comment (calvinrp) [reply crates/wasm-pkg-loader/src/main.rs]: @lann I think we have a path forward on the config, from our discussions. As for this PR and this config, how should we proceed?

Comment (lann) [reply crates/wasm-pkg-loader/src/main.rs]: Can we just revert this chunk of the PR?

Comment (calvinrp) [reply crates/wasm-pkg-loader/src/main.rs]: @lann Reverted
